src/metabase/lib/drill_thru.cljc

```diff
@@ -0,0 +1,612 @@
+(ns metabase.lib.drill-thru
+  (:require
+   [metabase.lib.aggregation :as lib.aggregation]
+   [metabase.lib.binning :as lib.binning]
+   [metabase.lib.breakout :as lib.breakout]
+   [metabase.lib.equality :as lib.equality]
+   [metabase.lib.filter :as lib.filter]
+   [metabase.lib.filter.operator :as lib.filter.operator]
+   [metabase.lib.metadata :as lib.metadata]
+   [metabase.lib.metadata.calculation :as lib.metadata.calculation]
+   [metabase.lib.options :as lib.options]
+   [metabase.lib.order-by :as lib.order-by]
+   [metabase.lib.ref :as lib.ref]
+   [metabase.lib.schema :as lib.schema]
+   [metabase.lib.schema.drill-thru :as lib.schema.drill-thru]
+   [metabase.lib.schema.expression :as lib.schema.expression]
+   [metabase.lib.temporal-bucket :as lib.temporal-bucket]
+   [metabase.lib.types.isa :as lib.types.isa]
+   [metabase.lib.util :as lib.util]
+   [metabase.util.malli :as mu]))
+
+;; TODO: Different ways to apply drill-thru to a query.
+;; So far:
+;; - :filter on each :operators of :drill-thru/quick-filter applied with (lib/filter query stage filter-clause)
+
+;; TODO: ActionMode, PublicMode, MetabotMode need to be captured in the FE before calling `available-drill-thrus`.
+
+;; TODO: Directly clicking on a table's primary key opens the object detail, with no popup? How does that work?
+
+;;; ---------------------------------------- Internals -------------------------------------------
+(defn- structured-query? [query stage-number]
+  (-> (lib.util/query-stage query stage-number)
+      :lib/type
+      (= :mbql.stage/mbql)))
+
+(defn- drill-thru-dispatch [_query _stage-number drill-thru]
+  (:type drill-thru))
```

Comment:
There are tons of implementations of `drill-thru-method` that are written to take > 3 args